File: frontend/app/intellij-plugin/src/main/kotlin/io/mateu/ijp/plugin/AppRegistry.kt
package io.mateu.ijp.plugin

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.intellij.ide.plugins.PluginManagerCore
import com.intellij.openapi.application.ApplicationInfo
import com.intellij.openapi.extensions.PluginId
import java.net.URI
import java.net.URLEncoder
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.nio.charset.StandardCharsets
import java.time.Duration

/**
 * App registry client — the same renderer-agnostic contract the React Native renderer boots
 * through: the installable carries only a registry URL + app id, and the registry entry maps the
 * app id to the Mateu base URL, the launch parameters (seeded into appState) and the required
 * renderer versions. For this renderer the relevant block is `intellij`:
 *
 *   {
 *     "baseUrl": "https://apps.example.com/admin",
 *     "parameters": { "tenantId": "1111" },
 *     "intellij": {
 *       "requiredPluginVersion": "0.2.0",
 *       "requiredIdeBuild": "243.0",
 *       "downloadUrl": "https://plugins.example.com/mateu-latest.zip"
 *     }
 *   }
 *
 * The entry URL is `{registryUrl}/{appId}.json`, or the registry URL itself with an `{appId}`
 * placeholder substituted.
 */
object AppRegistry {

    const val PLUGIN_ID = "io.mateu.ijp.plugin"

    data class Entry(
        val baseUrl: String,
        val parameters: Map<String, Any?>,
        val requiredPluginVersion: String?,
        val requiredIdeBuild: String?,
        val downloadUrl: String?,
    )

    /** A version requirement the installed plugin/IDE does not meet; null when boot may proceed. */
    data class Block(val message: String, val downloadUrl: String?)

    fun entryUrl(registryUrl: String, appId: String): String {
        val encoded = URLEncoder.encode(appId, StandardCharsets.UTF_8)
        if (registryUrl.contains("{appId}")) return registryUrl.replace("{appId}", encoded)
        return registryUrl.trimEnd('/') + "/" + encoded + ".json"
    }

    fun fetch(registryUrl: String, appId: String, mapper: ObjectMapper): Entry {
        val url = entryUrl(registryUrl, appId)
        val client = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(10)).build()
        val request = HttpRequest.newBuilder(URI.create(url))
            .header("Accept", "application/json")
            .timeout(Duration.ofSeconds(15))
            .GET()
            .build()
        val response = client.send(request, HttpResponse.BodyHandlers.ofString())
        if (response.statusCode() !in 200..299) {
            throw IllegalStateException("Registry answered HTTP ${response.statusCode()} for $url")
        }
        val node = mapper.readTree(response.body())
        val baseUrl = node.path("baseUrl").asText("")
        if (baseUrl.isBlank()) throw IllegalStateException("Registry entry for '$appId' has no baseUrl")
        val intellij = node.path("intellij")
        return Entry(
            baseUrl = baseUrl.trimEnd('/'),
            parameters = node.path("parameters").toMap(mapper),
            requiredPluginVersion = intellij.path("requiredPluginVersion").textOrNull(),
            requiredIdeBuild = intellij.path("requiredIdeBuild").textOrNull(),
            downloadUrl = intellij.path("downloadUrl").textOrNull(),
        )
    }

    /** Numeric dotted-version compare; missing/non-numeric segments count as 0 ("1.2" == "1.2.0"). */
    fun compareVersions(a: String, b: String): Int {
        val pa = a.split('.')
        val pb = b.split('.')
        for (i in 0 until maxOf(pa.size, pb.size)) {
            val na = pa.getOrNull(i)?.takeWhile { it.isDigit() }?.toIntOrNull() ?: 0
            val nb = pb.getOrNull(i)?.takeWhile { it.isDigit() }?.toIntOrNull() ?: 0
            if (na != nb) return na - nb
        }
        return 0
    }

    fun installedPluginVersion(): String =
        runCatching { PluginManagerCore.getPlugin(PluginId.getId(PLUGIN_ID))?.version }
            .getOrNull() ?: "0.0.0"

    /** The IDE build ("243.22562.145"); null outside a running IDE (renderProbe). */
    fun installedIdeBuild(): String? =
        runCatching { ApplicationInfo.getInstance().build.asStringWithoutProductCode() }.getOrNull()

    /** Checks the entry's version requirements against the installed plugin and IDE. */
    fun versionBlock(entry: Entry): Block? {
        val requiredPlugin = entry.requiredPluginVersion
        if (requiredPlugin != null && compareVersions(installedPluginVersion(), requiredPlugin) < 0) {
            return Block(
                "This app needs Mateu plugin $requiredPlugin — you have ${installedPluginVersion()}.",
                entry.downloadUrl,
            )
        }
        val requiredIde = entry.requiredIdeBuild
        val ide = installedIdeBuild()
        if (requiredIde != null && ide != null && compareVersions(ide, requiredIde) < 0) {
            return Block(
                "This app needs IntelliJ build $requiredIde or newer — you are on $ide.",
                null,
            )
        }
        return null
    }

    private fun JsonNode.textOrNull(): String? = if (isMissingNode || isNull) null else asText().ifBlank { null }

    private fun JsonNode.toMap(mapper: ObjectMapper): Map<String, Any?> =
        if (isObject) {
            @Suppress("UNCHECKED_CAST")
            mapper.convertValue(this, Map::class.java) as Map<String, Any?>
        } else {
            emptyMap()
        }
}

File: frontend/app/react-native/src/core/AppRegistry.ts
import Constants from 'expo-constants';

/**
 * App registry: the INSTALLABLE only carries a registry URL and an app id; a public registry
 * maps the app id to the Mateu base URL, the launch parameters (seeded into appState) and the
 * renderer version the app requires. Retargeting a backend or raising the required version is
 * then a registry edit — no republishing.
 *
 * Registry contract (renderer-agnostic): a GET on the entry URL returns
 *
 *   {
 *     "appId": "demo-admin-panel",
 *     "baseUrl": "https://apps.example.com/admin",
 *     "parameters": { "tenantId": "1111" },
 *     "requiredRendererVersion": "1.2.0",
 *     "storeUrl": { "android": "https://play.google.com/…", "ios": "https://apps.apple.com/…" }
 *   }
 *
 * The entry URL is `{registryUrl}/{appId}.json` (so any static hosting works), or the registry
 * URL itself with the `{appId}` placeholder substituted (for dynamic registries).
 */

export interface RegistryEntry {
  appId?: string;
  baseUrl: string;
  /** Seeded into the appState sent with every request (like @AppContext values). */
  parameters?: Record<string, unknown>;
  /** Minimum renderer (installable) version able to run this app. */
  requiredRendererVersion?: string;
  /** Store fallback when the renderer cannot self-update over the air. */
  storeUrl?: { android?: string; ios?: string };
}

export interface RegistryConfig {
  registryUrl: string;
  appId: string;
}

/** The installable's registry coordinates: `expo.extra` in app.json (production builds) or the
 *  EXPO_PUBLIC_MATEU_REGISTRY_URL / EXPO_PUBLIC_MATEU_APP_ID env vars (dev). Null → no registry,
 *  the renderer boots against its dev config. */
export function readRegistryConfig(): RegistryConfig | null {
  const extra = (Constants.expoConfig?.extra ?? {}) as Record<string, unknown>;
  const registryUrl =
    (extra['mateuRegistryUrl'] as string) || process.env.EXPO_PUBLIC_MATEU_REGISTRY_URL || '';
  const appId = (extra['mateuAppId'] as string) || process.env.EXPO_PUBLIC_MATEU_APP_ID || '';
  return registryUrl && appId ? { registryUrl, appId } : null;
}

export function registryEntryUrl({ registryUrl, appId }: RegistryConfig): string {
  if (registryUrl.includes('{appId}')) return registryUrl.replace('{appId}', encodeURIComponent(appId));
  return `${registryUrl.replace(/\/+$/, '')}/${encodeURIComponent(appId)}.json`;
}

export async function fetchRegistryEntry(config: RegistryConfig): Promise<RegistryEntry> {
  const url = registryEntryUrl(config);
  const response = await fetch(url, { headers: { Accept: 'application/json' } });
  if (!response.ok) throw new Error(`Registry answered HTTP ${response.status} for ${url}`);
  const entry = (await response.json()) as RegistryEntry;
  if (!entry || typeof entry.baseUrl !== 'string' || !entry.baseUrl) {
    throw new Error(`Registry entry for '${config.appId}' has no baseUrl`);
  }
  return entry;
}

/** Numeric dotted-version compare: negative when a < b, 0 when equal, positive when a > b.
 *  Non-numeric segments compare as 0, missing segments as 0 ("1.2" == "1.2.0"). */
export function compareVersions(a: string, b: string): number {
  const pa = a.split('.');
  const pb = b.split('.');
  for (let i = 0; i < Math.max(pa.length, pb.length); i++) {
    const na = parseInt(pa[i] ?? '0', 10) || 0;
    const nb = parseInt(pb[i] ?? '0', 10) || 0;
    if (na !== nb) return na - nb;
  }
  return 0;
}

/** The installed renderer version — the installable's `expo.version`. */
export function installedRendererVersion(): string {
  return Constants.expoConfig?.version ?? '0.0.0';
}

export function updateRequired(entry: RegistryEntry): boolean {
  const required = entry.requiredRendererVersion;
  if (!required) return false;
  return compareVersions(installedRendererVersion(), required) < 0;
}
